modules/internal/iam-validator/main.tf

```diff
@@ -0,0 +1,121 @@
+# Copyright 2025 "Google LLC"
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Data source that runs the permission check script in-line.
+data "external" "gcp_permission_check" {
+  program = [
+    "/bin/bash",
+    "-c",
+    # The script is updated to output a clean list of missing permissions on failure.
+    <<-EOT
+      set -e
+      if [ "$#" -ne 2 ]; then
+        echo "Usage: $0 <project_id> <permissions_comma_separated>" >&2; exit 1
+      fi
+
+      PROJECT_ID="$1"
+      PERMISSIONS_TO_CHECK="$2"
+
+      TOKEN=$(gcloud auth print-access-token)
+
+      PERMS_JSON_ARRAY=""
+      while IFS= read -r p; do
+        if [ -z "$PERMS_JSON_ARRAY" ]; then
+          PERMS_JSON_ARRAY="\"$p\""
+        else
+          PERMS_JSON_ARRAY="$PERMS_JSON_ARRAY,\"$p\""
+        fi
+      done < <(echo "$PERMISSIONS_TO_CHECK" | tr ',' '\n')
+      REQUEST_BODY="{\"permissions\":[$PERMS_JSON_ARRAY]}"
+
+      API_RESPONSE=$(curl -s -X POST \
+        -H "Authorization: Bearer $TOKEN" \
+        -H "Content-Type: application/json" \
+        -d "$REQUEST_BODY" \
+        "https://cloudresourcemanager.googleapis.com/v1/projects/$PROJECT_ID:testIamPermissions")
+
+      MISSING_PERMS=""
+      while IFS= read -r p; do
+        if ! echo "$API_RESPONSE" | grep -q "\"$p\""; then
+          if [ -z "$MISSING_PERMS" ]; then
+            MISSING_PERMS="$p"
+          else
+            MISSING_PERMS="$MISSING_PERMS,$p"
+          fi
+        fi
+      done < <(echo "$PERMISSIONS_TO_CHECK" | tr ',' '\n')
+
+      if [ -z "$MISSING_PERMS" ]; then
+        echo '{"validation_passed": "true", "missing_permissions_list": ""}'
+      else
+        # On failure, output the missing permissions in a clean, comma-separated list.
+        printf '{"validation_passed": "false", "missing_permissions_list": "%s"}' "$MISSING_PERMS"
+      fi
+    EOT
+    ,
+    "permission_checker.sh",
+    var.project_id,
+    join(",", var.required_permissions)
+  ]
+}
+
+locals {
+  # A map to suggest roles for missing permissions.
+  permission_to_role_map = {
```

Comment:
One suggestion here: Instead of hard coding this mapping, we can get the mapping info through terraform API: https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/iam_role

This adds to the complexity. So, doing this as part of another PR works as well.